doc/pekwm.1
.nh
.TH pekwm 1 "February 2021" pekwm "User Manual"

.SH NAME
.PP
pekwm \- a tabbed window manager


.SH SYNOPSIS
.PP
pekwm [\-\-display]


.SH DESCRIPTION
.PP
pekwm is a tabbed window manager which focuses on being highly
configurable. pekwm has a rich feature set including:

.RS
.IP \(bu 2
Autoproperties, set properties and group windows automatically based on type or title.
.IP \(bu 2
Keychains, multi\-level keybindings.
.IP \(bu 2
XRandR and Xinerama support.

.RE

.PP
For more information about the pekwm project and for full
documentation, visit http://www.pekwm.se.


.SH OPTIONS
.PP
\fB\-\-help\fP Show help information.

.PP
\fB\-\-version\fP Show version information.

.PP
\fB\-\-info\fP Show information about compiled in features.

.PP
\fB\-\-display\fP \fIDISPLAY\fP Connect to DISPLAY instead of DISPLAY set in environment.

.PP
\fB\-\-config\fP \fICONFIG\fP Use CONFIG file instead of default ~/.pekwm/config

.PP
\fB\-\-replace\fP Replace running window manager.


NEWS
NEWS pekwm-0.1.7
================

Bugs fixed

  * #XX Several memory leaks identified with valgrind plugged.
  * #XX Fixed broken iconfied state reported by
     <giuliani{.}blackfenix{@}gmail{.}com>
  * #XX CfgParser parsed data twice.
  * #XX StatusWindow placement fixed.
  * #XX Fix for _NET_WM_NAME ops by Florian Hoppe <flp{@}floppi{.}de>
  * #22 Font type is now possible to define as the first argument in
     themes.
  * #30 Panel raise issues, reported and fixed by Jyri Jokinen
     <shared{@}adresh{.}com>
  * #21 Dynamic menu entries become unselectable after first
     visit. Reported by Lassi Pölönen and fixed by Andreas
     <ioerror{@}lavabit{.}com>
  * #14 SendKey action requested by nidan.
  * #29 drag'n'drop requires the recieving window to be moved before
     working. Reported by Jyri Jokinen <shared{@}adresh{.}com>, fixed
     by Andreas <ioerror{@}lavabit{.}com>
  * #32 Shaping issues after shading. Reported by Paulo Igor Barra
     Nascimento <pibarnas{@}yahoo{.}com{.}br> and fixed by Andreas
     <ioerror{@}lavabit{.}com>
  * #XX Corrects the interpretation of the motif windowmanager hints
     by Andreas <ioerror{@}lavabit{.}com>
  * #XX Remove perl scripts reducing dependencies on packages. 

User visible changes

  * UTF-8 support added for reading and rendering window titles.
  * _NET_WM_ICON is now read and rendered in menus.
  * RandR support for multi-screen configurations.
  * New default theme.
  * CmdDialog size is now calculated from the current head and not the
    whole display.

Configuration changes

  * _NET_WM_WINDOW_TYPE handling configurable via autoproperties.
  * Width/Height 0 in autoproperties now means screen width/height.
  * WorkspacesPerRow and WorkspaceNames added with matching
    GotoWorkspace/SendToWorkspace parameters Up, PrevV, Down, NextV.
  * ShowWorkspaceIndicator
  * WorkspaceIndicatorScale
  * PCRE support remove.
  * EdgeIndent option added to Screen section of main configuration
    file. If true reserves space where the screen edge is.
  * EdgeSize now supports per edge configuration. 

Theme changes

  * Shaping suport for non-corner windows.
  * XMB type is used by default in favour of X11 type fonts.
  * Font alpha is now handles as % from 0-100 instead of previous Xft
    unsigned int value.
  * HeightAdapt option added to theme.
  * Font type is now possible to define as the first argument in themes.


src/pekwm.cc
//
// pekwm.cc for pekwm
// Copyright (C) 2021 Claes Nästén <pekdon@gmail.com>
//
// This program is licensed under the GNU GPL.
// See the LICENSE file for more information.
//

#include "Compat.hh"

#include <iostream>
#include <vector>

extern "C" {
#include <sys/types.h>
#include <sys/wait.h>
#include <errno.h>
#include <string.h>
#include <signal.h>
#include <unistd.h>
}

static int child_signal = 0;

static void
sigHandler(int signal)
{
    switch (signal) {
    case SIGHUP:
    case SIGINT:
    case SIGTERM:
        child_signal = signal;
        break;
    }
}

static int
waitPid(pid_t pid)
{
    int ret;
    while (waitpid(pid, &ret, 0) == -1 && errno == EINTR) {
        if (child_signal) {
            kill(pid, child_signal);
            child_signal = 0;
        }
    }

    if (WIFEXITED(ret)) {
        return WEXITSTATUS(ret);
    }
    return ret;
}

static bool
promptForRestart(const std::string& pekwm_dialog)
{
    pid_t pid = fork();
    if (pid == 0) {
        char *argv[] = {strdup(pekwm_dialog.c_str()),
                        strdup("-o"), strdup("Restart"),
                        strdup("-o"), strdup("Exit"),
                        strdup("-t"), strdup("pekwm crashed!"),
                        strdup("-r"),
                        strdup("pekwm quit unexpectedly, restart?"),
                        NULL};
        execvp(argv[0], argv);

        exit(1);
    } else if (pid == -1) {
        return false;
    } else {
        return waitPid(pid) == 0;
    }
}

static int
handleOkResult(char *argv[], int read_fd)
{
    char buf[1024] = {0};
    read(read_fd, buf, sizeof(buf) - 1);
    close(read_fd);

    if (strncmp("stop", buf, 4) == 0) {
        return 0;
    }
    if (strncmp("error", buf, 5) == 0) {
        return 1;
    }

    if (strncmp("restart ", buf, 8) == 0) {
        auto command = std::string(buf + 8);

        if (command.empty()) {
            execvp(argv[0], argv);
        } else {
            command = "exec " + command;
            execl("/bin/sh", "sh" , "-c", command.c_str(), (char*) 0);
        }

        std::cerr << "failed to run restart command: " << command << std::endl;
    }

    return 1;
}


static int
handleUnexpectedResult(char *argv[], int read_fd)
{
    close(read_fd);

    // run pekwm_dialog and wait for answer on to restart or not
    auto pekwm_dialog = std::string(argv[0]) + "_dialog";
    if (promptForRestart(pekwm_dialog)) {
        execvp(argv[0], argv);
        std::cerr << "failed to restart pekwm" << std::endl;
        exit(1);
    }

    std::cerr << "not restarting after crash" << std::endl;

    return 1;
}

/**
 * Main function of pekwm
 */
int
main(int argc, char *argv[])
{
    setenv("PEKWM_ETC_PATH", SYSCONFDIR, 1);
    setenv("PEKWM_SCRIPT_PATH", DATADIR "/pekwm/scripts", 1);
    setenv("PEKWM_THEME_PATH", DATADIR "/pekwm/themes", 1);

    // Get the pekwm_wm command by appending _wm to the path to ensure
    // the correct pekwm_wm is used when running from a non-installed
    // location.
    auto pekwm_wm = std::string(argv[0]) + "_wm";

    // Setup environment, DISPLAY is set to make RestartOther work as
    // expected re-using the display from the --display argument.
    //
    // PEKWM_CONFIG_FILE is set as an environment to make pekwm_dialog
    // catch the correct configuration file.
    std::vector<std::string> wm_argv = { pekwm_wm };
    for (int i = 1; i < argc; i++) {
        if ((strcmp("--display", argv[i]) == 0) && ((i + 1) < argc)) {
            setenv("DISPLAY", argv[++i], 1);
        } else if ((strcmp("--config", argv[i]) == 0) && ((i + 1) < argc)) {
            setenv("PEKWM_CONFIG_FILE", argv[++i], 1);
        } else {
            wm_argv.push_back(argv[i]);
        }
    }

    // Run the window manager inside a child process to avoid stopping
    // the X11 server on a crash, instead just restart
    int fd[2];
    if (pipe(fd) == -1) {
        std::cerr << "Failed to create pipe for communicating with "
                  << " pekwm_wm process" << std::endl;
        return 1;
    }

    int ret = 1;
    pid_t pid = fork();
    if (pid == 0) {
        // close reading end on child
        close(fd[0]);

        wm_argv.insert(wm_argv.begin() + 1, "--fd");
        wm_argv.insert(wm_argv.begin() + 2, std::to_string(fd[1]));

        int argc = 0;
        auto c_wm_argv = new char*[wm_argv.size() + 1];
        for (auto it : wm_argv) {
            c_wm_argv[argc++] = strdup(it.c_str());
        }
        c_wm_argv[argc++] = NULL;

        execvp(c_wm_argv[0], c_wm_argv);

        std::cerr << "Failed to execute: " << pekwm_wm << std::endl;
        exit(1);
    } else if (pid == -1) {
        std::cerr << "Failed to fork: " << strerror(errno) << std::endl;
    } else {
        // close writing end on parent
        close(fd[1]);

        struct sigaction act;
        act.sa_handler = sigHandler;
        act.sa_mask = sigset_t();
        act.sa_flags = SA_NOCLDSTOP | SA_NODEFER;
        sigaction(SIGHUP, &act, 0);
        sigaction(SIGTERM, &act, 0);
        sigaction(SIGINT, &act, 0);

        ret = waitPid(pid);
        if (ret == 0) {
            ret = handleOkResult(argv, fd[0]);
        } else {
            ret = handleUnexpectedResult(argv, fd[0]);
        }
    }

    return ret;
}
